### PR TITLE
Simplify `d2` computation in `apply_tof_kernel_and_symm_transformation`

### PR DESCRIPTION
**Pull request by robbietuk**
Modifiying the `d2` computation to remove the additional `inner_product` computations. The "mean difference"  between the old `d2` and new `d2` computations is ~< 1e-5. I assume this is a round error. 

Additionally, precompute the length `diff_unit_vector` before for loop.

_____
Profile of `apply_tof_kernel_and_symm_transformation` in a forward projection:
- `tof_sino_UCL`: 61.6k samples
- this PR: 49.1k samples (25% faster)

___
TODO:
- [x] Test

### CODE REVIEW

**Comment by robbietuk** [comment]
@KrisThielemans @NikEfth @ALEXJAZZ008008, thoughts and any suggestions on the tests?

P.S. in this test, using a slightly weird scanner geometry, this for loop is iterated ~218 million times. See scanner
[t2.txt](https://github.com/NikEfth/STIR/files/8007126/t2.txt)

File: src/include/stir/recon_buildblock/ProjMatrixByBin.inl [hunk 1]
```diff
@@ -149,24 +148,21 @@ ProjMatrixByBin::apply_tof_kernel_and_symm_transformation(ProjMatrixElemsForOneB
...
 
-    const float lor_length = 1.f / (std::sqrt(diff.x() * diff.x() +
+    const float lor_length = std::sqrt(diff.x() * diff.x() +
                                         diff.y() * diff.y() +
-                                        diff.z() * diff.z()));
+                                        diff.z() * diff.z());
+
+    const CartesianCoordinate3D<float> diff_unit_vector(diff.z()/lor_length,
+                                                        diff.y()/lor_length,
+                                                        diff.x()/lor_length);
```
Comment (KrisThielemans) [inline src/include/stir/recon_buildblock/ProjMatrixByBin.inl]: replace all this with
```C++
const CartesianCoordinate3D<float> diff_unit_vector(diff/norm(diff));
```

File: src/include/stir/recon_buildblock/ProjMatrixByBin.inl [hunk 2]
```diff
@@ -149,24 +148,21 @@ ProjMatrixByBin::apply_tof_kernel_and_symm_transformation(ProjMatrixElemsForOneB
...
 
-        voxel_center =
-                image_info_sptr->get_physical_coordinates_for_indices (c);
-
-        project_point_on_a_line(point1, point2, voxel_center);
-
-        const CartesianCoordinate3D<float> x = voxel_center - middle;
-
-        const float d2 = -inner_product(x, diff) * lor_length;
+        const float d2 = -inner_product(image_info_sptr->get_physical_coordinates_for_indices (c) - middle, diff_unit_vector);
```
Comment (KrisThielemans) [inline src/include/stir/recon_buildblock/ProjMatrixByBin.inl:159]: could precompute `-inner_product( - middle, diff_unit_vector)` outside of the loop, which would remove 2 subtractions. On the other hand, it is less clear, so I wouldn't (unless very well documented).

**Review by KrisThielemans — CHANGES_REQUESTED** [review]
As this is according to my comments, I'm happy 😄 . It makes the code much more understandable.

I can't see what tests we would write for this modification. It's pure geometry, and clearer from the original. This whole calculation should be tested by the TOF tests anyway.

**Comment by KrisThielemans** [comment]
2 appveyor buids fail because of time-out. ignore

File: src/include/stir/recon_buildblock/ProjMatrixByBin.inl (hunk 2 above)
Comment (robbietuk) [reply src/include/stir/recon_buildblock/ProjMatrixByBin.inl:159]: This is tough and I think we would need to rename a number of variable to do this clearly. I can do that but it does change the PR a fair bit. 

I think precomputing the `inner_product`, outside the for loop, would be ideal. The majority of the computational cost for TOF forward projection is still in this method. Any improvements to reduce repeated unnessisery computation would be ideal. I will give it another look today.

Comment (KrisThielemans) [reply src/include/stir/recon_buildblock/ProjMatrixByBin.inl:159]: > This is tough and I think we would need to rename a number of variable to do this clearly. I can do that but it does change the PR a fair bit.

I suggest we don't do it. Fast but unclear code is in the end unproductive.



> I think precomputing the `inner_product`, outside the for loop, would be ideal. 

obviously, we need an `inner_product` as it depends on the voxel.

> The majority of the computational cost for TOF forward projection is still in this method. Any improvements to reduce repeated unnessisery computation would be ideal.

as I've said elsewhere, I think writing this in terms of a TOF kernel (not relying on `erf`) I think is ultimately faster **and** clearer. That'd remove multiplications with `* r_sqrt2_gauss_sigma` and also the tests on `> 4` (which are repeated in the `erf` calculation anyway.

Ultimately though, whatever we do here is wasteful. It's better to invest time in the symmetries PR #9 such that this calculation will be cached and not repeated. Or indeed use `ParallelProj`

Comment (robbietuk) [reply src/include/stir/recon_buildblock/ProjMatrixByBin.inl:159]: > obviously, we need an inner_product as it depends on the voxel.


Sorry, early morning and I wasnt thinking...

>Ultimately though, whatever we do here is wasteful. It's better to invest time in the symmetries PR https://github.com/NikEfth/STIR/pull/9 such that this calculation will be cached and not repeated. Or indeed use ParallelProj

Okay. I think we merge this change now as it is a simplification. Then lets finalise the erfMapping/interpolation in https://github.com/robbietuk/STIR/pull/23 (not a PR on NikEfth:tof_sino_ucl yet) and I'll start to look into #9.

**Comment by robbietuk** [comment]
`test_time_of_flight` passes. Ready to merge?

**Review by KrisThielemans — APPROVED** [review]

**Comment by KrisThielemans** [comment]
@NikEfth could you squash-merge this please? (please edit the commit message to something short like "simplify calculations based on geometry, and move as much as possible of the for loop. This provides a moderate speed-up."

**Comment by KrisThielemans** [comment]
@robbietuk can you squash the 3 commits and force-push this here? Then I'll merge to `tof_sino_UCL`

**Comment by robbietuk** [comment]
Done but we may want to wait for the tests.

**Comment by robbietuk** [comment]
I believe that Appveyor failed due to `C:\projects\stir\src\include\stir/recon_buildblock/ProjMatrixByBin.inl(151): warning C4244: 'argument': conversion from 'double' to 'const float', possible loss of data [...]`. 

I think this relates to the use of `norm(diff)`, where diff is `const CartesianCoordinate3D<float>` and `norm(diff)` returns a double...
______
Static cast solution?
```
const CartesianCoordinate3D<float> diff_unit_vector(diff/static_cast<float>(norm(diff)));
```

**Comment by KrisThielemans** [comment]
Appveyor will not fail because of warnings (we didn't tell it to). Instead, the first 2 jobs are going just above the 1hr time limit.

Of course, having less warnings will decrease the job time somewhat, but just fixing this one won't help with that problem.  But it's still a good idea to put this in of course.

**Comment by robbietuk** [comment]
An easy fix to implement so force-pushed. Looks good. Ready to merge?

**Pull request merged**